packages/common/src/doc-cache.ts

```diff
@@ -0,0 +1,129 @@
+import { LRUMap } from 'lru_map'
+import DocID from "@ceramicnetwork/docid"
+
+import { DocStateHolder } from "./doctype"
+
+export type OnEvictFunction = (doc: DocStateHolder) => Promise<void>
+
+/**
+ * Encapsulates document caching (base, commits)
+ */
+export class DocCache {
+    private readonly _cacheCommits: boolean
+
+    private readonly _baseDocCache: LRUMap<string, DocStateHolder>
+    private readonly _commitDocCache: LRUMap<string, DocStateHolder>
+    private readonly _pinnedDocCache: Record<string, DocStateHolder>
+
+    constructor(limit, onEvictFn: OnEvictFunction, cacheCommits = true) {
+        this._cacheCommits = cacheCommits
+        this._baseDocCache = this._initDocLRUCache(limit, onEvictFn)
+        this._pinnedDocCache = {}
+
+        // use the same 'limit' if cacheCommits is enabled
+        this._commitDocCache = this._cacheCommits? this._initDocLRUCache(limit, onEvictFn) : new LRUMap(0)
+    }
+
+    /**
+     * Initialize single LRU cache
+     * @private
+     */
+    _initDocLRUCache(limit: number, onEvictFn: OnEvictFunction) {
+        const cache = new LRUMap<string, DocStateHolder>(limit)
+        if (onEvictFn) {
+            cache.shift = function() {
+                const entryArr = LRUMap.prototype.shift.call(cache)
+                if (entryArr) {
+                    onEvictFn(entryArr[1]) // TODO handle async call
+                }
+                return entryArr
+            }
+        }
+        return cache
+    }
+
+    /**
+     * Puts to cache
+     * @param doc - DocStateHolder instance
+     * @param isPinned - Is document pinned?
+     */
+    put(doc: DocStateHolder, isPinned = false): void {
```

Comment:
Can we remove the default value for this?  It seems useful to force developers to always have to explicitly think about whether the doc is pinned or not before putting it into the cache.

EDIT: Seeing how this is used, it seems to only ever be called with `isPinned` as true as part of the explicit pinning/unpinning workflow.  Could we actually separate this to a separate method, rather than an arg to `put`?  We could have a `pin` and an `unpin` method on the cache directly.  I think that makes it clearer what's happening here, and would also address my comment below about the behavior of `del` by making it clear that `unpin` doesn't actually delete a document from the cache, it just moves that document to the regular part of the cache where it is subject to eventual eviction down the road.